async_class.py

```diff
@@ -135,31 +162,71 @@ def create_future(self) -> asyncio.Future:
     async def __adel__(self) -> None:
         pass
 
+    def __init_subclass__(cls, **kwargs: KwargsType):
+        if getattr(cls, "__await__") is not AsyncObject.__await__:
+            raise TypeError("__await__ redeclaration is forbidden")
+
+    def __await__(self) -> Generator[Any, None, "AsyncObject"]:
+        if not hasattr(self, "_async_class_task_store"):
+            self._async_class_task_store = TaskStore(self.loop)
+
+        yield from self.create_task(
+            self.__ainit__(*self._args, **self._kwargs),
+        ).__await__()
+        return self
+
     def __del__(self) -> None:
         if self.__closed:
             return
 
-        self.close()
+        with suppress(BaseException):
+            self.loop.create_task(self.close())
 
-    def close(self) -> asyncio.Future:
+    async def close(self, exc: Optional[Exception] = None) -> None:
         if self.__closed:
-            f = asyncio.get_event_loop().create_future()
-            f.set_result(True)
-            return f
+            return
 
-        self.__closed = True
+        tasks: List[Union[asyncio.Future, Coroutine]] = []
+
+        if hasattr(self, "_async_class_task_store"):
+            tasks.append(self.__adel__())
+            tasks.append(self.__tasks__.close(exc))
+            self.__closed = True
```

Comment:
but here in close u switch closed flag. Code is reachable cuz you call await instance and task store exists here